nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp

```diff
@@ -0,0 +1,310 @@
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      logger_,
+      "BinaryFilter: Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  if (msg->type != BINARY_FILTER) {
+    RCLCPP_ERROR(logger_, "BinaryFilter: Mode %i is not supported", msg->type);
+    return;
+  }
+
+  mask_topic_ = msg->filter_mask_topic;
+
+  // Setting new filter mask subscriber
+  RCLCPP_INFO(
+    logger_,
+    "BinaryFilter: Subscribing to \"%s\" topic for filter mask...",
+    mask_topic_.c_str());
+  mask_sub_ = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    mask_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&BinaryFilter::maskCallback, this, std::placeholders::_1));
+}
+
+void BinaryFilter::maskCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    RCLCPP_INFO(
+      logger_,
+      "BinaryFilter: Received filter mask from %s topic.", mask_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "BinaryFilter: New filter mask arrived from %s topic. Updating old filter mask.",
+      mask_topic_.c_str());
+    filter_mask_.reset();
+  }
+
+  filter_mask_ = msg;
+  mask_frame_ = msg->header.frame_id;
+}
+
+bool BinaryFilter::transformPose(
+  const geometry_msgs::msg::Pose2D & pose,
+  geometry_msgs::msg::Pose2D & mask_pose) const
+{
+  if (mask_frame_ != global_frame_) {
+    // Filter mask and current layer are in different frames:
+    // Transform (pose.x, pose.y) point from current layer frame (global_frame_)
+    // to mask_pose point in mask_frame_
+    geometry_msgs::msg::TransformStamped transform;
+    geometry_msgs::msg::PointStamped in, out;
+    in.header.stamp = clock_->now();
+    in.header.frame_id = global_frame_;
+    in.point.x = pose.x;
+    in.point.y = pose.y;
+    in.point.z = 0;
+
+    try {
+      tf_->transform(in, out, mask_frame_, transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        logger_,
+        "BinaryFilter: failed to get costmap frame (%s) "
+        "transformation to mask frame (%s) with error: %s",
+        global_frame_.c_str(), mask_frame_.c_str(), ex.what());
+      return false;
+    }
+    mask_pose.x = out.point.x;
+    mask_pose.y = out.point.y;
+  } else {
+    // Filter mask and current layer are in the same frame:
+    // Just use pose coordinates
+    mask_pose = pose;
+  }
+
+  return true;
+}
+
+bool BinaryFilter::worldToMask(double wx, double wy, unsigned int & mx, unsigned int & my) const
```

Comment:
Same here -- if there are any of these functions that are reimplemented in multiple filters that might be helpful for future implementation, we might want to abstract them a layer down. This seems like another good candidate

Reply:
Done in https://github.com/ros-planning/navigation2/pull/3228/commits/9fba2a9bff21d33455daf83e24142179552d03ec